From 3383322321c41ed7d5faff24a98417cf4b27a75e Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Thu, 29 Aug 2019 12:10:32 +0800
Subject: [PATCH 32/81] HACK: modesetting: exa: Fix memory leak of screen
 pixmap

The exaCloseScreen() would be called before miCloseScreen(), so we need
to destroy the mi screen pixmap earlier.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 hw/xfree86/drivers/modesetting/driver.c | 9 +++++++++
 mi/miscrinit.c                          | 5 ++++-
 2 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/hw/xfree86/drivers/modesetting/driver.c b/hw/xfree86/drivers/modesetting/driver.c
index 4438a6f24..c671695a0 100644
--- a/hw/xfree86/drivers/modesetting/driver.c
+++ b/hw/xfree86/drivers/modesetting/driver.c
@@ -2224,6 +2224,15 @@ CloseScreen(ScreenPtr pScreen)
     }
 
     if (ms->drmmode.exa) {
+        PixmapPtr screen_pixmap;
+
+        /* Destroy exa screen pixmap before deinit exa */
+        screen_pixmap = pScreen->GetScreenPixmap(pScreen);
+        if (screen_pixmap == pScreen->devPrivate) {
+            pScreen->DestroyPixmap(screen_pixmap);
+            pScreen->devPrivate = NULL;
+        }
+
         ms_deinit_exa(pScrn);
     }
 
diff --git a/mi/miscrinit.c b/mi/miscrinit.c
index d66ac417a..8e8db58d6 100644
--- a/mi/miscrinit.c
+++ b/mi/miscrinit.c
@@ -134,7 +134,10 @@ miCloseScreen(ScreenPtr pScreen)
 {
     miScreenPtr pScreenPriv = miGetScreenPriv(pScreen);
 
-    ((*pScreen->DestroyPixmap) ((PixmapPtr) pScreen->devPrivate));
+    if (pScreen->devPrivate) {
+        ((*pScreen->DestroyPixmap) ((PixmapPtr) pScreen->devPrivate));
+        pScreen->devPrivate = NULL;
+    }
 
     pScreen->CloseScreen = pScreenPriv->CloseScreen;
 
-- 
2.38.5

